video/out/gpu/error_diffusion.c

```diff
@@ -141,54 +136,85 @@ void pass_error_diffusion(struct gl_shader_cache *sc,
 
     // The dithering will quantize pixel value into multiples of 1/dither_quant.
     int dither_quant = (1 << depth) - 1;
-    // The absolute value of the errors to propagate is less than 1/dither_quant,
-    // multiply by dither_quant24 to have them processed with int in 24 bit
-    // precision.
-    double dither_quant24 = (double)(1 << 24) * dither_quant;
+
+    // We encode errors in RGB components into a single 32-bit unsigned integer.
+    // The error we propagate from the current pixel is in range of
+    // [-0.5 / dither_quant, 0.5 / dither_quant]. While not quite obvious, the
+    // sum of all errors been propagated into a pixel is also in the same range.
+    // It's possible to map errors in this range into [-127, 127], and use an
+    // unsigned 8-bit integer to store it (using standard two's complement).
+    // The three 8-bit unsigned integers can then be encoded into a single
+    // 32-bit unsigned integer, with two 4-bit padding to prevent addition
+    // operation overflows affecting other component. There are at most 12
+    // addition operations on each pixel, so 4-bit padding should be enough.
+    // The following figure is how the encoding looks like.
+    //
+    //     +------------------------------------+
+    //     |RRRRRRRR|0000|GGGGGGGG|0000|BBBBBBBB|
+    //     +------------------------------------+
+    //
```

Comment:
What happens when we overflow addition on the `R` component?

Reply:
The high bit will be discarded, if there is an overflow after add operation. This is actually intended, to have operations on signed and unsigned integers align to each other.

    (-1)+(-1)=0xffffffff+0xffffffff=0xfffffffe=-2
    (-1)*(-1)=0xffffffff+0xffffffff=0x00000001=1